### ISSUE TITLE
GBA GameShark Code Converter Doesn't Work

### ISSUE THREAD

**Issue by darkgob** [TASVideos/BizHawk#2545]
### Summary
GBA GameShark codes are not parsed properly in the Code Converter, resulting in an error.

### Repro
1. Tools -> Cheats -> Code Converter
2. Enter a GBA GameShark code in the proper format (8 characters, a space, then 8 more characters; 17 characters total).
3. Click Convert button.

### Output
```
All GBA GameShark Codes need to be 17 characters in length with a space after the first eight.
```

### Fix

- File: GbaGameSharkDecoder.cs
  - Function: Decode(string code)
    - [Line 60](https://github.com/TASVideos/BizHawk/blob/c1a734a04df3f3914133028629d1734ad36b71d7/src/BizHawk.Client.Common/cheats/GbaGameSharkDecoder.cs#L60): Index of space character is expected to be 8, not 9 (zero-indexed)
    - [Line 79](https://github.com/TASVideos/BizHawk/blob/c1a734a04df3f3914133028629d1734ad36b71d7/src/BizHawk.Client.Common/cheats/GbaGameSharkDecoder.cs#L79): Second parameter to Substring() should be 7, not 8
    - [Line 79 (again)](https://github.com/TASVideos/BizHawk/blob/c1a734a04df3f3914133028629d1734ad36b71d7/src/BizHawk.Client.Common/cheats/GbaGameSharkDecoder.cs#L79): Need to use NumberStyles.HexNumber as second parameter to int.Parse() (as was done on Line 78)

### Host env.
- BizHawk 2.5.2; Win10 Pro 1909; AMD/AMD
- BizHawk 2.5.3; OS Unknown; Hardware Unknown
- BizHawk dev build at 8199e6777; Win10 Home 1909; AMD/AMD

**Comment by Sappharad** [comment]
It's not that simple. There were multiple GBA devices called "Gameshark" made by different developers and they used different code formats.

The GameShark brand was owned by Interact and they originally licensed the Action Replay from Datel UK as GameShark. They had this relationship for number of years, from the PS1 / N64 era all the way to PS2/GBA era. All GameShark devices produced during this period were nearly identical to their Action Replay counterparts sold in the UK. During the early PS2/GBA era, Datel decided to no longer license their technology to InterAct (or some other situation) which resulted in Datel setting up their own US offices in Florida and officially selling the Action Replay in  North America. At this point, InterAct decided to continue the Gameshark brand, so they partnered with Pelican out of Europe to rebrand their Xploder cheat devices as GameShark in the US. The later model Gameshark used an entirely different 12 character code format.  (The story is a bit more complex than that if you consider that prior to GameShark being a rebranded Xploder, Xploder devices were licensed for the US market as CodeBreaker. The team behind that product actually ended up producing their own without Pelican later. The original devices prior to that are quite rare now and go over over 3x their original price on Ebay.)

After the breakup, Datel released GBA Action Replay v3, which changed their code own format and encryption, so that owners of the legacy GameShark products would essentially lose official support for their devices.

What you're referencing here is the legacy GameShark, prior to the split between Datel and InterAct. While it's true that the official formatting contains a space and it's how those codes should be displayed, it is purely a formatting choice and the whitespace itself has no value when it comes to the actual code data.

---
tl;dr - You should just ignore the whitespace altogether, since it doesn't matter.

**Comment by darkgob** [comment]
> tl;dr - You should just ignore the whitespace altogether, since it doesn't matter.

Not really sure why you went into any of that as it's not really relevant to the problem.

1. The code as currently written expects a space to be in the code string at index 9, but the error message suggests to the user that the expected format has the space at index 8.
2. The index parameters to Substring() are similarly wrong as this ends up including the space in the resulting substring.
3. Even after fixing the above problem, int.Parse() would still fail because it can't parse a string representing a hexadecimal number without the NumberStyles.HexNumber parameter.

If you feel that the space should not be required at all, that's fine but it's not relevant here.

**Comment by Sappharad** [comment]
> If you feel that the space should not be required at all, that's fine but it's not relevant here.

Not at all. As long as there are the correct number of hex characters to decode the code, it makes sense to accept it.
I was just rambling about the complexity that is GBA cheat devices, because there were at least 6 different dedicated cheat devices for the GBA, not even counting flashcarts that had cheat support.

**Comment by darkgob** [comment]
> > If you feel that the space should not be required at all, that's fine but it's not relevant here.
> 
> Not at all. As long as there are the correct number of hex characters to decode the code, it makes sense to accept it. (I agree with you) I was just rambling about the complexity that is GBA cheat devices, because there were at least 6 different dedicated cheat devices for the GBA, not even counting flashcarts that had cheat support. It was worth mentioning since there is more than 1 code format for GBA Gameshark depending when the device was made.

According to the history of this class (which requires some digging to find since the folder structure was changed in May), the code design of how GameShark codes are handled was significantly overhauled back in February (in the before-times, in the long-long ago), at which time it was stated in the commit notes that the previous GBA GameShark code was a "mess" that was just starting to be detangled.

Given that, it should really be treated as a separate design issue.  As it currently stands, "eight-space-eight" codes are intended to be the only supported format for GBA GameShark codes in BizHawk, but even those don't work right now; the 3 fixes proposed in this issue resolve that.

**Comment by YoshiRulz** [comment]
This sounds like an easy fix but I'd rather write tests first. Can someone give me a diverse sample of GBA codes? (For other consoles too!)

**Comment by Sappharad** [comment]
> Can someone give me a diverse sample of GBA codes? (For other consoles too!)

http://bsfree.org/
On GBA specifically you’ll want the version 1 and 2 section. This is just for Gameshark devices.
